src/AsyncRequest.php

```diff
@@ -0,0 +1,255 @@
+<?php
+
+namespace Onoi\HttpRequest;
+
+use InvalidArgumentException;
+use Closure;
+
+/**
+ * This class creates a remote socked connection to initiate an asynchronous http
+ * request.
+ *
+ * Once a connection is established and content has been posted, the request
+ * is expected to close the connection. The receiving client is expected to open
+ * a separate process and initiated a independent transaction from that of the
+ * originating request.
+ *
+ * A callback can verify the acknowledge but not the transactional response
+ * (as it would be for a cURL request).
+ *
+ * @license GNU GPL v2+
+ * @since 1.1
+ *
+ * @author mwjames
+ */
+class AsyncRequest implements HttpRequest {
+
+	/**
+	 * @var array
+	 */
+	private $options = array();
+
+	/**
+	 * @var integer
+	 */
+	private $errno = 0;
+
+	/**
+	 * @var string
+	 */
+	private $errstr = '';
+
+	/**
+	 * @var string
+	 */
+	private $lastTransferInfo = '';
+
+	/**
+	 * @var Closure|null
+	 */
+	private $callback = null;
+
+	/**
+	 * @since 1.1
+	 *
+	 * @param string|null $url
+	 */
+	public function __construct( $url = null ) {
+		$this->setOption( ONOI_HTTP_REQUEST_URL, $url );
+		$this->setOption( ONOI_HTTP_REQUEST_CONNECTION_TIMEOUT, 15 );
+		$this->setOption( ONOI_HTTP_REQUEST_CONNECTION_FAILURE_REPEAT, 2 );
+		$this->setOption( ONOI_HTTP_REQUEST_STREAM_CLIENT_FLAGS, STREAM_CLIENT_ASYNC_CONNECT | STREAM_CLIENT_CONNECT );
+		$this->setOption( ONOI_HTTP_REQUEST_METHOD, 'POST' );
+		$this->setOption( ONOI_HTTP_REQUEST_CONTENT_TYPE, "application/x-www-form-urlencoded" );
+		$this->setOption( ONOI_HTTP_REQUEST_SSL_VERIFYPEER, false );
+	}
+
+	/**
+	 * @since 1.1
+	 *
+	 * {@inheritDoc}
+	 */
+	public function ping() {
+
+		if ( $this->getOption( ONOI_HTTP_REQUEST_URL ) === null ) {
+			return false;
+		}
+
+		$urlComponents = $this->getUrlComponents( $this->getOption( ONOI_HTTP_REQUEST_URL ) );
+
+		$resource = $this->getResourceFromSocketClient(
+			$urlComponents,
+			STREAM_CLIENT_CONNECT
+		);
+
+		if ( !$resource ) {
+			return false;
+		}
+
+		stream_set_timeout( $resource, $this->getOption( ONOI_HTTP_REQUEST_CONNECTION_TIMEOUT ) );
+
+		$httpMessage = (
+			"HEAD "  . $urlComponents['path'] . " HTTP/1.1\r\n" .
+			"Host: " . $urlComponents['host'] . "\r\n" .
+			"Connection: Close\r\n\r\n"
+		);
+
+		$res = @fwrite( $resource, $httpMessage );
+		@fclose( $resource );
+
+		return $res !== false;
+	}
+
+	/**
+	 * @since 1.1
+	 *
+	 * @param Closure $callback
+	 */
+	public function setCallback( Closure $callback ) {
```

Comment:
Instead of cluttering the class with different setter, the following is now applied.

```
		$this->httpRequest->setOption( ONOI_HTTP_REQUEST_ON_COMPLETED_CALLBACK, function( $requestResponse ) {
			wfDebugLog( 'smw', 'SMW\AsyncJobDispatchManager: ' . json_encode( $requestResponse->getList() ) . "\n" );
		} );

		$this->httpRequest->setOption( ONOI_HTTP_REQUEST_ON_FAILED_CALLBACK, function( $requestResponse ) use ( $dispatchableCallbackJob, $title, $type, $parameters ) {
			wfDebugLog( 'smw', "SMW\AsyncJobDispatchManager: Connection to SpecialAsyncJobDispatcher failed therefore adding {$type} for " . $title->getPrefixedDBkey() . "\n" );
			call_user_func_array( $dispatchableCallbackJob, array( $title, $parameters ) );
		} );
```